File: src/main/java/org/eclipse/sonarttcn3/measures/ModuleMetrics.java
/******************************************************************************
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html
 ******************************************************************************/
package org.eclipse.sonarttcn3.measures;

import java.util.List;
import java.util.Map;
import static java.util.Map.entry;

import org.sonar.api.measures.Metric;
import org.sonar.api.measures.Metrics;

public class ModuleMetrics implements Metrics {
    public static final Metric<Integer> NOF_STATEMENTS =
        new Metric.Builder("NOF_STATEMENTS", "Number of statements", Metric.ValueType.INT)
            .setDescription("Number of statements in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> NOF_ALTSTEPS =
        new Metric.Builder("NOF_ALTSTEPS", "Number of altsteps", Metric.ValueType.INT)
            .setDescription("Number of altsteps in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> NOF_TESTCASES =
        new Metric.Builder("NOF_TESTCASES", "Number of test cases", Metric.ValueType.INT)
            .setDescription("Number of test cases in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> IN_ENVY =
        new Metric.Builder("IN_ENVY", "Internal feature envy", Metric.ValueType.INT)
            .setDescription("Number of references to entities inside the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> OUT_ENVY =
        new Metric.Builder("OUT_ENVY", "External feature envy", Metric.ValueType.INT)
            .setDescription("Number of references to entities outside the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();
    
    public static final Metric<Integer> NOF_FUNCTIONS =
        new Metric.Builder("NOF_FUNCTIONS", "Number of functions", Metric.ValueType.INT)
            .setDescription("Number of functions in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> NOF_FIXME =
        new Metric.Builder("NOF_FIXME", "Fixme comments", Metric.ValueType.INT)
            .setDescription("Number of comments beginning with \"FIXME\"")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> TIMES_IMPORTED =
        new Metric.Builder("TIMES_IMPORTED", "Imported count", Metric.ValueType.INT)
            .setDescription("Times the module was imported by other modules")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> NOF_IMPORTS =
        new Metric.Builder("NOF_IMPORTS", "Number of imports", Metric.ValueType.INT)
            .setDescription("Count of module importations in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> EFFERENT_COUPLING =
        new Metric.Builder("EFFERENT_COUPLING", "External assignment references", Metric.ValueType.INT)
            .setDescription("Number of referred assignments that are defined outside the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> AFFERENT_COUPLING =
        new Metric.Builder("AFFERENT_COUPLING", "Internal assignment references", Metric.ValueType.INT)
            .setDescription("Number of referred assignments that are defined inside the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> INSTABILITY =
        new Metric.Builder("INSTABILITY", "Efferent to (efferent plus afferent) coupling ratio", Metric.ValueType.INT)
            .setDescription("Efferent to (efferent plus afferent) coupling ratio")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    public static final Metric<Integer> LINES_OF_CODE =
        new Metric.Builder("LINES_OF_CODE", "Lines of code", Metric.ValueType.INT)
            .setDescription("Number of code lines in the module")
            .setDirection(Metric.DIRECTION_BETTER)
            .setDomain("Ttcn3Plugin")
            .setQualitative(false)
            .create();

    static Map<String, Metric<Integer>> moduleMetrics = Map.ofEntries(
        entry("NOF_STATEMENTS", NOF_STATEMENTS),
        entry("NOF_ALTSTEPS", NOF_ALTSTEPS),
        entry("NOF_TESTCASES", NOF_TESTCASES),
        entry("IN_ENVY", IN_ENVY),
        entry("OUT_ENVY", OUT_ENVY),
        entry("NOF_FUNCTIONS", NOF_FUNCTIONS),
        entry("NOF_FIXME", NOF_FIXME),
        entry("TIMES_IMPORTED", TIMES_IMPORTED),
        entry("NOF_IMPORTS", NOF_IMPORTS),
        entry("EFFERENT_COUPLING", EFFERENT_COUPLING),
        entry("AFFERENT_COUPLING", AFFERENT_COUPLING),
        entry("INSTABILITY", INSTABILITY),
        entry("LINES_OF_CODE", LINES_OF_CODE)
    );

    @Override
    public List<Metric> getMetrics() {
        return List.copyOf(moduleMetrics.values());
    }

    public static Metric<Integer> getMetric(final String key) {
        return moduleMetrics.get(key);
    }
}

File: src/main/resources/static/module_metrics.js
const uiPageSize = 10;
let uiCurrentPage = 1;
let totalPages = 0;
let allComponents = [];
let handlers = false;

const metrics = new Map([
    [ 'NOF_STATEMENTS', 'Number of statements' ],
    [ 'NOF_ALTSTEPS', 'Number of altsteps' ],
    [ 'NOF_TESTCASES', 'Number of test cases' ],
    [ 'IN_ENVY', 'Internal feature envy' ],
    [ 'OUT_ENVY', 'External feature envy' ],
    [ 'NOF_FUNCTIONS', 'Number of functions' ],
    
    // This is broken in Titanium right now
    // [ 'NOF_FIXME', 'Fixme comments' ],
    
    [ 'TIMES_IMPORTED', 'Times imported' ],
    [ 'NOF_IMPORTS', 'Number of imports' ],
    [ 'EFFERENT_COUPLING', 'External assignment references' ],
    [ 'AFFERENT_COUPLING', 'Internal assignment references' ],
    [ 'INSTABILITY', 'Efferent to (efferent plus afferent) coupling ratio' ],
    [ 'LINES_OF_CODE', 'Lines of code' ]
]);

const container = document.createElement('div');
container.style.padding = '12px';
container.textAlign = 'center';

const titleContainer = document.createElement('div');
const title = document.createElement('h3');
title.textContent = 'TTCN3 Module Metrics';
title.style.marginBottom = '10px';
container.appendChild(titleContainer);
const buttonBack = document.createElement('button');
buttonBack.paddingLeft = '5px';
setButtonStyle(buttonBack);
buttonBack.textContent = ' < ';
const buttonForward = document.createElement('button');
setButtonStyle(buttonForward);
buttonForward.textContent = ' > ';
const pageText = document.createTextNode('');
const buttonContainer = document.createElement('span');
buttonContainer.appendChild(buttonBack);
buttonContainer.appendChild(pageText);
buttonContainer.appendChild(buttonForward);
container.appendChild(titleContainer);

const table = document.createElement('table');
table.style.borderCollapse = 'collapse';
table.style.width = '800px';
table.style.maxWidth = '100%';
table.style.fontFamily = 'sans-serif';
table.style.overflow = 'auto';
table.style.flex = '1';
const thead = document.createElement('thead');
const headerRow = document.createElement('tr');

[ 'Path / metric ', 'Value' ].forEach((text, index) => {
    const th = document.createElement('th');
    if (index === 0) {
        const thWithButtons = document.createElement('span');
        const label = document.createTextNode(text);
        label.paddingLeft = '5px';
        thWithButtons.appendChild(label);
        thWithButtons.appendChild(buttonContainer);
        th.appendChild(thWithButtons);
    } else {
        th.textContent = text;
    }
    th.style.borderBottom = '2px solid #ccc';
    th.style.textAlign = 'left';
    th.style.padding = '8px';
    th.style.background = '#f5f5f5';
    th.style.zIndex = '100';
    th.style.top = '0';
    th.style.position = 'sticky';
    headerRow.appendChild(th);
});

thead.appendChild(headerRow);
table.appendChild(thead);
const tbody = document.createElement('tbody');
table.appendChild(tbody);
const tableContainer = document.createElement('div');
tableContainer.style.overflow = 'auto';
tableContainer.style.flex = '1';
tableContainer.style.display = 'flex';
tableContainer.style.flexDirection = 'column';
tableContainer.style.alignItems = 'center';
tableContainer.style.height = '70vh';
tableContainer.style.borderBottom = '2px solid #000';
tableContainer.appendChild(table);
container.appendChild(tableContainer);

const metricDescriptions = new Map([
    [ 'NOF_STATEMENTS', 'Number of statements in the module' ],
    [ 'NOF_ALTSTEPS', 'Number of altsteps in the module' ],
    [ 'NOF_TESTCASES', 'Number of test cases in the module' ],
    [ 'IN_ENVY', 'Number of references to entities inside the module' ],
    [ 'OUT_ENVY', 'Number of references to entities outside the module' ],
    [ 'NOF_FUNCTIONS', 'Number of functions in the module' ],

    // This is broken in Titanium right now
    // [ 'NOF_FIXME', 'Number of comments beginning with "FIXME"' ],
    
    [ 'TIMES_IMPORTED', 'Times the module was imported by other modules' ],
    [ 'NOF_IMPORTS', 'Count of module importations in the module' ],
    [ 'EFFERENT_COUPLING', 'Number of referred assignments that are defined outside the module' ],
    [ 'AFFERENT_COUPLING', 'Number of referred assignments that are defined inside the module' ],
    [ 'INSTABILITY', 'Efferent to (efferent plus afferent) coupling ratio' ],
    [ 'LINES_OF_CODE', 'Number of code lines in the module' ]
].sort((a, b) => a < b));

function getProjectKey() {
    const params = new URLSearchParams(window.location.search);
    return params.get('id') || params.get('project');
}

async function loadMetrics() {
    let p = 1;
    const ps = 100;
    const projectKey = getProjectKey();
    let count = 0;

    do {
        const res = await fetch(
            `/api/measures/component_tree?component=${projectKey}&metricKeys=${[...metrics.keys()].join(',')}&qualifiers=FIL&p=${p}&ps=${ps}`
        );

        const data = await res.json();
        allComponents = allComponents.concat(
            data.components.filter(comp => comp.language === 'ttcn3')
        );

        total = data.paging.total;
        count += ps;
        p++;
    } while (count < total);
    totalPages = Math.ceil(total / uiPageSize);
}

function renderTable(components) {
    tbody.innerHTML = '';

    components.forEach(c => {
        const fileName = document.createElement('td');
        fileName.textContent = c.path;
        fileName.colSpan = 2;
        fileName.style.fontWeight = 'bold';
        fileName.style.paddingTop = '10px';
        const filerow = document.createElement('tr');
        filerow.appendChild(fileName);
        tbody.appendChild(filerow);
        if (c.measures.length === 0) {
            const row = document.createElement('tr');
            row.style.background = '#fff';
            row.style.paddingLeft = '10px';
            row.style.paddingTop = '3px';
            const measureCell = document.createElement('td');
            measureCell.colSpan = 2;
            measureCell.style.color = '#ff7a41';
            measureCell.textContent = 'No measures available';
            row.appendChild(measureCell);
            tbody.appendChild(row);
            return;
        }

        c.measures.forEach((measure, index) => {
            const value = measure.value || '0';
            const number = Number(value);
            if (isNaN(number)) {
                return;
            }

            const row = document.createElement('tr');
            row.style.background = index % 2 === 0 ? '#fff' : '#ded8d8';
            row.style.paddingLeft = '10px';
            row.style.paddingTop = '3px';

            const measureCell = document.createElement('td');
            measureCell.textContent = metrics.get(measure.metric);
            const measureDesc = metricDescriptions.get(measure.metric);
            if (measureDesc) {
                measureCell.title = measureDesc;
            }

            const valueCell = document.createElement('td');
            valueCell.textContent = value;
            valueCell.style.fontWeight = 'bold';

            row.appendChild(measureCell);
            row.appendChild(valueCell);

            tbody.appendChild(row);
        });
    });
    
    pageText.textContent = ` page ${uiCurrentPage}/${totalPages} `;
    return container.outerHTML;
}

function getComponentPage() {
    const first = (uiCurrentPage - 1) * uiPageSize;
    return allComponents.slice(first, first + uiPageSize);
}

function setButtonStyle(button) {
    button.style.background = "#0052CC";
    button.style.color = "white";
    button.style.border = "none";
    button.style.padding = "6px 12px";
    button.style.borderRadius = "4px";
    button.style.cursor = "pointer";

    button.onmouseenter = () => button.style.background = "#0065FF";
    button.onmouseleave = () => button.style.background = "#0052CC";
}

function setPagingButtonState() {
    buttonBack.disabled = uiCurrentPage === 1;
    buttonForward.disabled = uiCurrentPage === totalPages;
}

window.registerExtension('ttcn3/module_metrics', async function (options) {
    options.el.innerHTML = `<h2>Loading</h2>`;
    await loadMetrics();
    if (allComponents.length === 0) {
        options.el.innerHTML = `<h2>No measures available (missing analysis or non-TTCN3 project)</h2>`;
        return;
    }
    if (!handlers) {
        handlers = true;
        buttonForward.addEventListener('click', () => {
            if (uiCurrentPage < totalPages) {
                uiCurrentPage++;
                renderTable(getComponentPage() || []);        
            }
            setPagingButtonState();
        });
        buttonBack.addEventListener('click', () => {
            if (uiCurrentPage > 1) {
                uiCurrentPage--;
                renderTable(getComponentPage() || []);        
            }
            setPagingButtonState();
        });
    }
    renderTable(getComponentPage() || []);
    
    options.el.innerHTML = '';
    options.el.appendChild(container);
    return;
});
